### PR TITLE
Remove umpire and device cuda to a single source code file

### PR DESCRIPTION
**Pull request by koparasy**

### CODE REVIEW

File: src/AMSlib/AMS.cpp [hunk 1]
```diff
@@ -389,8 +389,8 @@ class AMSWrap
       } else if (log_prefix.find("<PID>") != std::string::npos) {
         pattern = std::string("<PID>");
         id = getpid();
-      }
-      // Combine hostname and pid
+      }  // Combine hostname and pid
```
Comment (ggeorgakoudis) [inline src/AMSlib/AMS.cpp]: Move comment above

File: src/AMSlib/ml/hdcache.hpp [hunk 2]
```diff
@@ -17,6 +17,7 @@
 #include <stdexcept>
 #include <string>
 #include <type_traits>
+#include <unordered_map>
```
Comment (ggeorgakoudis) [inline src/AMSlib/ml/hdcache.hpp:20]: unused?

File: src/AMSlib/wf/resource_manager.hpp [hunk 3]
```diff
@@ -205,13 +188,13 @@ class ResourceManager
...
                          size_t& as)
   {
-    RMAllocators[resource]->getAllocatorStats(wm, cs, as);
     return;
   }
 
   //! ------------------------------------------------------------------------
 };
 
+
```
Comment (ggeorgakoudis) [inline src/AMSlib/wf/resource_manager.hpp]: newline

File: tests/AMSlib/CMakeLists.txt [hunk 4]
```diff
@@ -155,8 +155,9 @@ function(ADDTEST exe test_name)
   endif()
 endfunction()
 
-BUILD_TEST(ams_allocator_test ams_allocate.cpp)
-ADDTEST(ams_allocator_test AMSAllocate)
+# This test requires Allocate
```
Comment (ggeorgakoudis) [inline tests/AMSlib/CMakeLists.txt:158]: dead code? are those tests going to be replaced?

File: src/AMSlib/wf/resource_manager.hpp [hunk 5]
```diff
@@ -9,54 +9,47 @@
...
+struct AMSAllocator;
+namespace internal
+{
+void _raw_copy(void* src,
+               AMSResourceType src_dev,
+               void* dest,
+               AMSResourceType dest_dev,
+               size_t num_bytes);
+
+AMSAllocator* _get_allocator(std::string alloc_name, AMSResourceType resource);
```
Comment (ggeorgakoudis) [inline src/AMSlib/wf/resource_manager.hpp]: std::string reference?

File: src/AMSlib/wf/resource_manager.hpp [hunk 6]
```diff
@@ -9,54 +9,47 @@
...
-        "in AMSAllocator(%d, %s, %p)",
-        id,
-        alloc_name.c_str(),
-        this)
-  }
 
-  ~AMSAllocator() { DBG(AMSAllocator, "in ~AMSAllocator(%d, %p)", id, this) }
+struct AMSAllocator {
+  std::string name;
+  AMSAllocator(std::string alloc_name) : name(alloc_name) {}
```
Comment (ggeorgakoudis) [inline src/AMSlib/wf/resource_manager.hpp]: std::string reference

File: src/AMSlib/wf/resource_manager.hpp [hunk 7]
```diff
@@ -9,54 +9,47 @@
...
+  virtual ~AMSAllocator() = default;
 
-  void* allocate(size_t num_bytes);
-  void deallocate(void* ptr);
+  virtual void* allocate(size_t num_bytes) = 0;
+  virtual void deallocate(void* ptr) = 0;
 
   std::string getName();
```
Comment (ggeorgakoudis) [inline src/AMSlib/wf/resource_manager.hpp]: `const std::string& getName() const`?

File: src/AMSlib/wf/resource_manager.hpp [hunk 8]
```diff
@@ -9,54 +9,47 @@
...
-  static void deregisterPtr(void* ptr)
-  {
-    auto& rm = umpire::ResourceManager::getInstance();
-    rm.deregisterAllocation(ptr);
-  }
-
-  void getAllocatorStats(size_t& wm, size_t& cs, size_t& as);
 };
 
+
```
Comment (ggeorgakoudis) [inline src/AMSlib/wf/resource_manager.hpp]: newline

File: src/AMSlib/wf/resource_manager.hpp [hunk 9]
```diff
@@ -116,37 +109,26 @@ class ResourceManager
...
-   *  @return void.
-   */
-  void deregisterExternal(void* ptr) { AMSAllocator::deregisterPtr(ptr); }
-
   /** @brief copy values from src to destination regardless of their memory location.
    *  @tparam TypeInValue type of pointers
    *  @param[in] src Source memory pointer.
    *  @param[out] dest destination memory pointer.
-   *  @param[in] size number of bytes to copy. (When 0 copies entire allocated area)
+   *  @param[in] size number of values to copy.
```
Comment (ggeorgakoudis) [inline src/AMSlib/wf/resource_manager.hpp]: does it always assume the value type does not contain nested pointers? add it to the comment?

File: src/AMSlib/wf/resource_manager.cpp [hunk 10]
```diff
@@ -5,48 +5,113 @@
  * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
  */
 
-#include "resource_manager.hpp"
-
+#include <cstdlib>
+#include <cstring>
 #include <umpire/ResourceManager.hpp>
 #include <umpire/Umpire.hpp>
 #include <umpire/strategy/QuickPool.hpp>
```
Comment (ggeorgakoudis) [inline src/AMSlib/wf/resource_manager.cpp]: remove?

File: src/AMSlib/wf/resource_manager.cpp [hunk 11]
```diff
@@ -5,48 +5,113 @@
...
-  rm.registerAllocation(ptr,
-                        umpire::util::AllocationRecord(
-                            ptr, nBytes, allocator.getAllocationStrategy()));
+void _raw_copy(void *src,
+               AMSResourceType src_dev,
+               void *dest,
+               AMSResourceType dest_dev,
+               size_t num_bytes)
+{
+  if (src_dev == AMSResourceType::AMS_HOST) {
```
Comment (ggeorgakoudis) [inline src/AMSlib/wf/resource_manager.cpp]: nested switch-case with fall-throughs to remove duplication for readability?

File: src/AMSlib/wf/resource_manager.cpp [hunk 12]
```diff
@@ -5,48 +5,113 @@
...
+  }
 }
 
-void AMSAllocator::getAllocatorStats(size_t &wm, size_t &cs, size_t &as)
+AMSAllocator *_get_allocator(std::string alloc_name, AMSResourceType resource)
 {
-  wm = allocator.getHighWatermark();
-  cs = allocator.getCurrentSize();
-  as = allocator.getActualSize();
+  if (resource == AMSResourceType::AMS_DEVICE) {
```
Comment (ggeorgakoudis) [inline src/AMSlib/wf/resource_manager.cpp]: switch-case for readability?

File: src/AMSlib/wf/resource_manager.cpp [hunk 13]
```diff
@@ -5,48 +5,113 @@
...
+    return ((num_to_round + multiple- 1) / multiple) * multiple;
 }
 
-void AMSAllocator::deallocate(void *ptr) { allocator.deallocate(ptr); }
+std::string AMSAllocator::getName() { return name; }
+
+
+struct AMSDefaultDeviceAllocator final : AMSAllocator {
+  AMSDefaultDeviceAllocator(std::string name) : AMSAllocator(name){};
+  ~AMSDefaultDeviceAllocator() = default;
```
Comment (ggeorgakoudis) [inline src/AMSlib/wf/resource_manager.cpp:29]: Shouldn't it also free/deallocate allocations in the destructor?

File: tests/AMSlib/CMakeLists.txt
Comment (ggeorgakoudis) [inline tests/AMSlib/CMakeLists.txt:1]: Remove umpire?

File: src/AMSlib/wf/device.hpp [hunk 14]
```diff
@@ -201,86 +203,124 @@ inline void HtoHMemcpy(void *dest, void *src, size_t nBytes)
...
-inline void HtoHMemcpy(void *dest, void *src, size_t nBytes)
+inline int pack(bool cond,
+                const bool *predicate,
+                const size_t n,
+                const TypeValue **sparse,
+                TypeValue **dense,
+                int dims)
 {
-  std::memcpy(dest, src, nBytes);
+  return -1;
```
Comment (ggeorgakoudis) [inline src/AMSlib/wf/device.hpp]: replace `-1` with a named constexpr or define

**Review by ggeorgakoudis — CHANGES_REQUESTED** [review]

File: src/AMSlib/AMS.cpp (hunk 1 above)
Comment (koparasy) [reply src/AMSlib/AMS.cpp]: Fixed.

File: src/AMSlib/ml/hdcache.hpp (hunk 2 above)
Comment (koparasy) [reply src/AMSlib/ml/hdcache.hpp:20]: We use an unordered map. It is likely it was pulling the header from another include.

File: src/AMSlib/wf/resource_manager.hpp (hunk 3 above)
Comment (koparasy) [reply src/AMSlib/wf/resource_manager.hpp]: Fixed.

File: tests/AMSlib/CMakeLists.txt (hunk 4 above)
Comment (koparasy) [reply tests/AMSlib/CMakeLists.txt:158]: Fixed.

File: src/AMSlib/wf/resource_manager.hpp (hunk 6 above)
Comment (koparasy) [reply src/AMSlib/wf/resource_manager.hpp]: Fixed.

File: src/AMSlib/wf/resource_manager.hpp (hunk 7 above)
Comment (koparasy) [reply src/AMSlib/wf/resource_manager.hpp]: Fixed.

File: src/AMSlib/wf/resource_manager.cpp (hunk 10 above)
Comment (koparasy) [reply src/AMSlib/wf/resource_manager.cpp]: Fixed.

File: src/AMSlib/wf/resource_manager.hpp (hunk 5 above)
Comment (koparasy) [reply src/AMSlib/wf/resource_manager.hpp]: Fixed.

File: src/AMSlib/wf/resource_manager.hpp (hunk 9 above)
Comment (koparasy) [reply src/AMSlib/wf/resource_manager.hpp]: Resolved.

File: src/AMSlib/wf/resource_manager.cpp (hunk 11 above)
Comment (koparasy) [reply src/AMSlib/wf/resource_manager.cpp]: Fixed.

File: src/AMSlib/wf/resource_manager.cpp (hunk 13 above)
Comment (koparasy) [reply src/AMSlib/wf/resource_manager.cpp:29]: We are not keeping track of allocations. It is the responsibility of the caller to deallocate data.

File: src/AMSlib/wf/resource_manager.cpp (hunk 12 above)
Comment (koparasy) [reply src/AMSlib/wf/resource_manager.cpp]: Fixed

File: src/AMSlib/wf/device.hpp (hunk 14 above)
Comment (koparasy) [reply src/AMSlib/wf/device.hpp]: Done.

File: tests/AMSlib/CMakeLists.txt
Comment (koparasy) [reply tests/AMSlib/CMakeLists.txt:1]: We will re-instate umpire behind this generic interface. I will leave it there.

**Review by ggeorgakoudis — APPROVED** [review]
LGTM

**Pull request merged**